File: tools/verify/render_text.c
/* C side for the text helpers — prints the same lines as render_text.js. */
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

#include "../../core/text.h"

#define BUF 512
#define MAXROWS 24
#define ROWCAP 64

int main(int argc, char **argv) {
  const char *path = argc > 1 ? argv[1] : "tools/verify/text.tsv";
  FILE *f = fopen(path, "r");
  if (!f) { perror(path); return 1; }

  char line[1024];
  while (fgets(line, sizeof line, f)) {
    if (line[0] == '/' || line[0] == '\n') continue;
    size_t len = strlen(line);
    while (len && (line[len - 1] == '\n' || line[len - 1] == '\r')) line[--len] = 0;

    char *fld[4], *p = line;
    int ok = 1;
    for (int i = 0; i < 4; i++) {
      char *t = strchr(p, '\t');
      if (!t) { ok = 0; break; }
      *t = 0; fld[i] = p; p = t + 1;
    }
    if (!ok) continue;

    const char *name = fld[0], *kind = fld[1];
    const int a = atoi(fld[2]), b = atoi(fld[3]);
    const char *text = p;

    char folded[BUF];
    deck_fold(text, folded, sizeof folded);

    if (strcmp(kind, "fold") == 0) {
      printf("%-18s |%s|\n", name, folded);

    } else if (strcmp(kind, "wrap") == 0) {
      static char rows[MAXROWS * ROWCAP];
      int n = deck_wrap(folded, a, rows, ROWCAP, MAXROWS);
      printf("%-18s |", name);
      for (int i = 0; i < n; i++)
        printf("%s%s", i ? "/" : "", rows + (size_t)i * ROWCAP);
      printf("|\n");

    } else if (strcmp(kind, "scroll") == 0) {
      deck_scroll_t sc = { 0, 0, 0.0 };
      char win[BUF];
      printf("%-18s |", name);
      for (int i = 0; i < b; i++) {
        deck_scroll(&sc, 100.0, folded, a, win, sizeof win);
        printf("%s%s", i ? "/" : "", win);
      }
      printf("|\n");
    }
  }
  fclose(f);
  return 0;
}

File: tools/verify/render_text.js
// Reference side for the text helpers.
//
// deckText, wrapLyric and scrollText live in legacy/web/app.js, which touches
// the DOM at load and cannot simply be evaluated. Rather than transcribe them
// (and risk verifying a copy against itself), cut the three function bodies out
// of the source by brace matching and run the originals.
const fs = require("fs");
const path = require("path");
const vm = require("vm");

const ROOT = path.join(__dirname, "..", "..");
const src = fs.readFileSync(path.join(ROOT, "legacy", "web", "app.js"), "utf8");

function extract(name) {
  const start = src.indexOf(`function ${name}(`);
  if (start < 0) throw new Error(`${name} not found in app.js`);
  let i = src.indexOf("{", start), depth = 0;
  for (let j = i; j < src.length; j++) {
    if (src[j] === "{") depth++;
    else if (src[j] === "}") { depth--; if (depth === 0) return src.slice(start, j + 1); }
  }
  throw new Error(`unbalanced braces in ${name}`);
}

const ctx = { LYRIC_CELLS: 30 };
ctx.globalThis = ctx;
vm.createContext(ctx);
vm.runInContext(
  [extract("deckText"), extract("wrapLyric"), extract("scrollText")].join("\n") +
  ";globalThis.__fold = deckText; globalThis.__wrap = wrapLyric;" +
  " globalThis.__scroll = scrollText;", ctx);

function out(name, s) { console.log(`${name.padEnd(18)} |${s}|`); }

const casesPath = process.argv[2] || path.join(ROOT, "tools", "verify", "text.tsv");
for (const raw of fs.readFileSync(casesPath, "utf8").split("\n")) {
  if (!raw || raw.startsWith("/")) continue;
  const line = raw.replace(/\r$/, "");
  const parts = line.split("\t");
  if (parts.length < 4) continue;
  const [name, kind, a, b] = parts;
  const text = parts.slice(4).join("\t");

  if (kind === "fold") {
    out(name, ctx.__fold(text));
  } else if (kind === "wrap") {
    const folded = ctx.__fold(text);
    out(name, ctx.__wrap(folded, Number(a)).join("/"));
  } else if (kind === "scroll") {
    // a = cells, b = number of 100 ms steps
    const folded = ctx.__fold(text);
    const sc = { offset: 0, phase: 0, t: 0 };
    const frames = [];
    for (let i = 0; i < Number(b); i++) frames.push(ctx.__scroll(sc, 100, folded, Number(a)));
    out(name, frames.join("/"));
  }
}
